homeassistant/components/matrix/__init__.py

```diff
@@ -444,16 +452,57 @@ async def _handle_multi_room_send(
             )
         await asyncio.wait(_tasks)
 
-    async def _send_image(
-        self, image_path: str, target_rooms: Sequence[RoomAnyID]
+    async def _send_image_url(
+        self, image_url: str, target_rooms: Sequence[RoomAnyID], verify_ssl: bool
     ) -> None:
-        """Upload an image, then send it to all target_rooms."""
-        _is_allowed_path = await self.hass.async_add_executor_job(
-            self.hass.config.is_allowed_path, image_path
+        """Upload an image from a URL, then send it to all target_rooms."""
+        # Get required image metadata.
+        if not self.hass.config.is_allowed_external_url(image_url):
+            _LOGGER.error("URL '%s' not in allow list", image_url)
+            return
+        session = aiohttp_client.async_get_clientsession(self.hass)
+        resp = await session.get(
+            image_url, raise_for_status=True, verify_ssl=verify_ssl
+        )
+        image_type = (
+            resp.headers.get("Content-Type")
+            or mimetypes.guess_type(image_url)[0]
+            or "image/jpeg"
         )
-        if not _is_allowed_path:
-            _LOGGER.error("Path not allowed: %s", image_path)
+        if (
+            int(resp.headers.get("Content-Length", "0"))
+            > MAX_ALLOWED_DOWNLOAD_SIZE_BYTES
+        ):
+            _LOGGER.error(
+                "Attachment too large (Content-Length reports %s). Max size: %s"
+                " bytes",
+                int(str(resp.headers.get("Content-Length"))),
+                MAX_ALLOWED_DOWNLOAD_SIZE_BYTES,
+            )
             return
+        async with aiofiles.tempfile.TemporaryDirectory() as tempdir:
+            image_extension = mimetypes.guess_extension(image_type)
+            path = pathlib.Path(tempdir) / f"image{image_extension}"
+            async with aiofiles.open(path, "wb") as f:
+                await f.write(await resp.read())
+            await self._send_image(
+                str(path.absolute()), target_rooms, skip_check_allowed=True
+            )
```

Comment:
Some sans/reduced-IO ideas:
- [Good] Use `AsyncSpooledTemporaryFile` instead; just like [SpooledTemporaryFile](https://docs.python.org/3/library/tempfile.html#tempfile.SpooledTemporaryFile), all the data is spooled in-memory instead of using the disk, which can be a precious resource on some low-resource systems.
- [Better] Ideally we don't even directly need the idea of a temporary file; we just need something to iteratively read bytes from. I think that modifying `_send_image` to accept either an image `Path` *or* `resp.read()` directly would easily cut out the need for temp files at all